### PR TITLE
[DUOS-1987][risk=no] Redirect to profile page if profile incomplete

### PR DESCRIPTION
**Pull request by connorlbark**
## Addresses

https://broadworkbench.atlassian.net/browse/DUOS-1987

Should redirect when:
- User who has not completed profile accepts ToS
- User who has accepted ToS but not completed profile logs in.

----
Have you read [Terra's Contributing Guide](https://github.com/DataBiosphere/terra-ui/wiki/Contributor-Guide) lately? If not, do that first.

- Label PR with a Jira ticket number and include a link to the ticket
- Label PR with a security risk modifier [no, low, medium, high]
- PR describes scope of changes
- Get a minimum of one thumbs worth of review, preferably two if enough team members are available
- Get PO sign-off for all non-trivial UI or workflow changes
- Verify all tests go green
- Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/libs/utils.js [hunk 1]
```diff
@@ -772,3 +772,12 @@ export const completedResearcherInfoCheck = (properties) => {
...
   return !isNil(institutionId);
 };
+
+export const researcherProfileComplete = (user) => {
+  const {
+    institutionId,
+    selectedSigningOfficialId,
+  } = getPropertyValuesFromUser(user);
+
+  return !isNil(institutionId) && !isNil(selectedSigningOfficialId);
```
Comment (connorlbark) [inline src/libs/utils.js]: Is this how we want to check if a profile is complete? Right now, checks that there is a valid institution id and selected sigining official id; in the case that the user suggests a signing official / institution, their profile is still considered incomplete; this seems right to me, but wanted to make sure.

Comment (connorlbark) [reply src/libs/utils.js]: thought: is the expected behavior for a user which is a signing official / any other non-researcher to always skip this ?

Comment (JVThomas) [reply src/libs/utils.js]: > Is this how we want to check if a profile is complete? Right now, checks that there is a valid institution id and selected sigining official id; in the case that the user suggests a signing official / institution, their profile is still considered incomplete; this seems right to me, but wanted to make sure.

I talked with @solideoglori about this and we both decided not to worry about a "complete" profile and only have the redirect occur on users that are registering with DUOS for the first time. Depending on the role you may not care about SO or Institution, so nil checks on both won't be enough (For example, DAC roles don't care about SO or Institution, SO requires an institution but not SO, and a research needs both SO and institution). At the same time the system allows for multiple roles which just adds complexity.

From a product perspective the odds of someone signing out of the profile page without doing anything is relatively low. If a user does do that, the only features that would require checks are the DAR Form and the SO console. The DAR page already performs that check and provides the user with a message and link to the profile page. For the SO we only assign that role if the user has an institution so we're okay there as well.

**TLDR**: Don't worry about a "complete" profile, just perform the redirect for new users. If the user already exists, just have the normal redirects occur.

File: src/pages/TermsOfServiceAcceptance.js [hunk 2]
```diff
@@ -23,6 +23,12 @@ export default function TermsOfServiceAcceptance(props) {
     const user = await Storage.getCurrentUser();
     await TosService.acceptTos();
     await Storage.setUserIsLogged(true);
+
+    if (user.isResearcher && !researcherProfileComplete(user)) {
+      history.push('/profile');
+      return;
+    }
```
Comment (rushtong) [inline src/pages/TermsOfServiceAcceptance.js]: I don't think we need to check for `user.isResearcher` here - all users are always researchers (that's our default user role).

Comment (rushtong) [reply src/pages/TermsOfServiceAcceptance.js]: I'm on the fence about the `researcherProfileComplete` check. I'm leaning to sending the user to the profile page whenever they accept ToS, regardless of completeness.

1. If a user has never selected an institution, both approaches work
2. If a user has previously filled in an institution, then rejected ToS, I think we want to send them to the profile page so they can see the on-boarding form again if they re-accept ToS. In this case, the current approach doesn't work.

File: src/components/SignIn.js [hunk 3]
```diff
@@ -42,9 +42,15 @@ export default function SignIn(props) {
     if (!isEmpty(userStatus) && !tosAccepted) {
       await Storage.setUserIsLogged(false);
       history.push('/tos_acceptance');
-    } else {
-      redirect(user);
+      return;
+    }
+
+    const {isResearcher} = Storage.getCurrentUser();
+    if (isResearcher && !researcherProfileComplete(user)) {
```
Comment (rushtong) [inline src/components/SignIn.js]: Same feedback here as below, `isResearcher` should be true for all users.

Comment (rushtong) [reply src/components/SignIn.js]: I think this is the right place for `researcherProfileComplete`

**Review by rushtong — COMMENTED** [review]
I'm proposing a change to the ToS redirect in comments below.

File: src/pages/TermsOfServiceAcceptance.js (hunk 2 above)
Comment (lu-c) [reply src/pages/TermsOfServiceAcceptance.js]: I don't have a strong opinion here, but I think when I use other websites, I would expect to be redirected to whatever my main home view would be when accepting the ToS. Like, when I've already had to do an extra unexpected click, I just want to see the content, and this extra profile step would be only mildly annoying (and hopefully only seen once :P)

**Comment by lu-c** [comment]
I have a few questions for clarity for this ticket:
1. Is the ToS the same thing as the onboarding form?
2. The ticket didnt describe anything about redirecting to the profile at all - was this addressed in a separate conversation?
3. Would you want to consider a localstorage approach? The ticket says "when a new user signs into duos", so I think it sounds like something that would make sense for tracking that ("`has the user logged in before` AND (`already seen the onboarding form` OR `profile completed`)? if no, redirect to onboarding form, if yes, then send them to the dataset catalog instead)

**Comment by rushtong** [comment]
> I have a few questions for clarity for this ticket:
> 
> 1. Is the ToS the same thing as the onboarding form?
> 2. The ticket didnt describe anything about redirecting to the profile at all - was this addressed in a separate conversation?

I can preemptively answer a couple of these

1. No, they're different things. The on-boarding form is on the profile page. The ToS is the page that shows up if you select "Terms of Service" in the footer, or have not accepted Terms of Service yet (or have accepted, then rejected ToS)
2. Since the onboarding form is on the profile page, that's why we're redirecting there.

**Comment by rushtong** [comment]
> 3\. ("`has the user logged in before` AND (`already seen the onboarding form` OR `profile completed`)

We can answer some of these pretty easily. When a user signs in, we make a couple of calls to the back end to figure out some details.

1. Has the user logged in before? If the user has accepted ToS, then we know they've logged in before.
2. Has the user seen the onboarding form? This is not clear. If the user has accepted ToS or if they've selected an institution, then we know that they have. However, we don't always know if they haven't (sorry for the double negative). If a user has accepted ToS, then rejected ToS, that user looks just like one who's never signed in before. 
3. Is the user's profile completed? That's an easy one - they've selected an institution. We should heed Justin's comment: 

> TLDR: Don't worry about a "complete" profile, just perform the redirect for new users.

**Comment by connorlbark** [comment]
Great discussion here!

Taking in feedback: removed the profile complete check on sign on, and always redirects to the profile page on ToS acceptance. This makes the ticket very simple - no more need for checking profile completeness, or any sort of logic.

**Review by rushtong — APPROVED** [review]
👍🏽

**Review by JVThomas — APPROVED** [review]
